drv/lpc55-sprot-server/src/handler.rs

```diff
@@ -568,7 +568,62 @@ impl<'a> Handler {
                 )?;
                 Ok((RspBody::Ok, None))
             }
+            ReqBody::State(StateReq::LifecycleState) => {
+                let out = self.lifecycle_state();
+
+                Ok((RspBody::State(out.map(StateRsp::LifecycleState)), None))
+            }
+        }
+    }
+
+    fn lifecycle_state(&mut self) -> Result<LifecycleState, StateError> {
+        const CMPA_SIZE: usize = 512;
+        let mut buf = [0u8; CMPA_SIZE];
+
+        // If the SHA-256 Digest is zeros, the CMPA is unlocked.  The SHA-256
+        // digest is located in the last 4 words of the CMPA.
+        self.update
+            .read_rot_page(RotPage::Cmpa, &mut buf)
+            .map_err(StateError::ReadCmpa)?;
+        if buf[CMPA_SIZE - 32..].iter().all(|b| *b == 0) {
+            // We use Unprogrammed as a catchall if the CMPA is unlocked
+            return Ok(LifecycleState::Unprogrammed);
         }
+
+        self.update
+            .read_rot_page(RotPage::CfpaActive, &mut buf)
+            .map_err(StateError::ReadCfpa)?;
+
+        // Look at the ROTKH_REVOKE byte
+        let revoke = buf[24];
+
+        // TODO use the type from dice_mfg_msgs?
+        enum S {
+            Invalid,
+            Enabled,
+            Revoked,
+        }
+        let slots = [0, 1, 2, 3].map(|i| match (revoke >> (i * 2)) & 0b11 {
+            0b00 => S::Invalid,
+            0b01 => S::Enabled,
+            _ => S::Revoked,
+        });
+
+        let state = match slots {
+            [S::Enabled, _, S::Invalid, S::Invalid]
+            | [_, S::Enabled, S::Invalid, S::Invalid] => {
+                LifecycleState::Release
+            }
+            [S::Revoked, S::Revoked, S::Enabled, _]
+            | [S::Revoked, S::Revoked, _, S::Enabled] => {
+                LifecycleState::Development
+            }
+            [S::Revoked, S::Revoked, S::Revoked, S::Revoked] => {
+                LifecycleState::EndOfLife
+            }
```

Comment:
This should be an impossible state. If we've revoked all keys we can't boot any secure boot code and we're locked per the above check so we can't even run this code.

Reply:
I think `unreachable!` is a better idea but if that would add in surprise panics we can go with a comment

Reply:
(we talked in chat, and are going with the comment to avoid a panic path)